File: apps/api/internal/models/location.go
package model

import (
	"database/sql"
	"time"
)

type Location struct {
	ID          string
	RegionID    string
	Name        string
	Description sql.NullString
	Notes       sql.NullString
	DmNotes     sql.NullString
	MapX        sql.NullFloat64
	MapY        sql.NullFloat64
	DeletedAt   sql.NullTime
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type CreateLocationRequest struct {
	RegionID    string
	CampaignID  string
	Name        string
	Description sql.NullString
	Notes       sql.NullString
	DmNotes     sql.NullString
	MapX        sql.NullFloat64
	MapY        sql.NullFloat64
}

type UpdateLocationRequest struct {
	ID          string
	CampaignID  string
	Name        *string
	Description sql.NullString
	Notes       sql.NullString
	DmNotes     sql.NullString
	MapX        sql.NullFloat64
	MapY        sql.NullFloat64
}

File: apps/web/src/features/regions/procedures/proto/location.ts
import { timestampDate } from "@bufbuild/protobuf/wkt";
import { ORPCError } from "@orpc/client";
import { LocationsSchema } from "@/features/regions/types";
import type { Location } from "@/gen/proto/planner/v1/location_pb";

export function protoToLocation(proto: Location) {
	if (!proto.createdAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "Location is missing createdAt",
		});
	}
	if (!proto.updatedAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "Location is missing updatedAt",
		});
	}

	return LocationsSchema.parse({
		createdAt: timestampDate(proto.createdAt),
		deletedAt: proto.deletedAt ? timestampDate(proto.deletedAt) : null,
		description: proto.description ?? null,
		dmNotes: proto.dmNotes ?? null,
		id: proto.id,
		mapX: proto.mapX ?? null,
		mapY: proto.mapY ?? null,
		name: proto.name,
		notes: proto.notes ?? null,
		regionId: proto.regionId,
		updatedAt: timestampDate(proto.updatedAt),
	});
}
